### PR TITLE
KAFKA-6928: Refactor StreamsPartitionAssignor retry logic

### PR DESCRIPTION
**Pull request by guozhangwang**
1. The retry loop of the InternalTopicManager would just be: a) describe topics, and exclude those which already exist with the right num.partitions, b) for the remaining topics, try to create them. Remove any inner loops.

2. In CreateTopicResponse and MetadataResponse (for describe topic), handle the special error code of TopicExist and UnknownTopicOrPartition in order to retry in the next loop.

3. Do not handle TimeoutException since it should already been handled inside AdminClient.

Add corresponding unit tests for a) topic marked for deletion but not complete yet, in which case metadata response would not contain this topic, but create topic would return error TopicExists; b) request keep getting timed out.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by guozhangwang** [comment]
@mjsax @bbejeck @vvcephei

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java [hunk 1]
```diff
@@ -95,173 +94,154 @@ public InternalTopicManager(final AdminClient adminClient,
...
                         log.error(INTERRUPTED_ERROR_MESSAGE, fatalException);
                         throw new IllegalStateException(INTERRUPTED_ERROR_MESSAGE, fatalException);
+                    } catch (final ExecutionException executionException) {
+                        final Throwable cause = executionException.getCause();
+                        if (cause instanceof TopicExistsException) {
+                            // This topic didn't exist earlier or its leader not known before; just retain it for next round of validation.
+                            log.info("Could not create topic {}. Topic is probably marked for deletion (number of partitions is unknown).\n" +
+                                "Will retry to create this topic in {} ms (to let broker finish async delete operation first).\n" +
+                                "Error message was: {}", topicName, retryBackOffMs, cause.toString());
+                        } else {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java:145]: What is the impact that we don't handle `TimeoutException` any longer? With internal admin client retries, could we still get a timeout exception?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java [hunk 2]
```diff
@@ -95,173 +94,154 @@ public InternalTopicManager(final AdminClient adminClient,
...
+        final Map<String, Integer> existedTopicPartition = new HashMap<>();
+        for (final Map.Entry<String, KafkaFuture<TopicDescription>> topicFuture : futures.entrySet()) {
+            final String topicName = topicFuture.getKey();
+            try {
+                final TopicDescription topicDescription = topicFuture.getValue().get();
+                existedTopicPartition.put(
+                    topicFuture.getKey(),
+                    topicDescription.partitions().size());
+            } catch (final InterruptedException fatalException) {
+                Thread.currentThread().interrupt();
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java:200]: nit: do you want to add a comment as above?

**Review by mjsax — COMMENTED** [review]
Overall LGTM.

Was wondering, if we should add an integration test? Ie, create repartition topic, delete it, start streams. It should hit this case eventually if delete is slow. Thoughts?

**Comment by guozhangwang** [comment]
> Was wondering, if we should add an integration test? Ie, create repartition topic, delete it, start streams. It should hit this case eventually if delete is slow. Thoughts?

I've thought about an integration test case, but the issue is that timing is quite hard to control here: most likely the delete topic will be completed even before streams starting phase start to ask for metadata. The unit test I've added here, is to verify that if a topic is marked for deletion but never completed, streams should "correctly" block in the retry logic until exhausted all retries and fail. I felt it provides a better coverage than the integration test itself.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java (hunk 1 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java:145]: `StreamsConfig.RETRIES_CONFIG` will be passed to AdminClient as `AdminClientConfig.RETRIES_CONFIG`; default value is 5.

Before this change, timeout exception will be retried up to 5 times in the outer loop, and with the inner admin client's loop of 5, it means that we would retry at most 5 * 5 = 25 times with backoff for timeouts.

After this change we only rely on inner admin client's loop; so technically speaking we are less resilient to timeouts, BUT I think the previous logic itself is just "more resilient by mistake" (i.e. a user bumping the num.retries from 5 to 6 would actually get 36 - 25 = 11 more retries...) and the current logic is better for users to understand: if you do observe that TimeoutException is being thrown, it means that the num.retries or retry.backoff is not sufficient and you should consider bump it up.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java (hunk 2 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java:200]: Ack.

**Comment by guozhangwang** [comment]
Cherry-pick to 2.1

**Pull request merged**